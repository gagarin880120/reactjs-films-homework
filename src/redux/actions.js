function searchResultsAction(results) {
  return {
      type: 'SEARCH',
      results: results
  };
}

function genresAction(genres) {
  return {
      type: 'GENRES',
      genres: genres
  };
}

export { searchResultsAction, genresAction };
