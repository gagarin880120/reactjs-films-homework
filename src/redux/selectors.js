import getGenresString from '../helpers/helpers';

const searchResultsSelector = (state) => {
  state.searchResults.map((v) => {
    const obj = v;
    obj.genres = getGenresString(v.genre_ids, state.genres);
    return { ...obj };
  });
  return state.searchResults;
};

const modalSelector = (state) => state.isModalOpen;

const trailerSelector = (state) => state.trailerURL;


export { searchResultsSelector, modalSelector, trailerSelector };
