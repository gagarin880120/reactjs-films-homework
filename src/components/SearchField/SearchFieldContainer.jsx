import React from 'react';
import { connect } from 'react-redux';
import { getResults } from '../../redux/actions';
import SearchField from './SearchField';

export function SearchFieldContainer({ onSearch }) {
  return (
    <SearchField
      onSearch={onSearch}
    />
  )
}

export const mapDispatchToProps = (dispatch) => {
  return {
    onSearch(query) {
      if(query) {
        dispatch(getResults(query));
      }
    }
  }
}

export default connect(null, mapDispatchToProps)(SearchFieldContainer)
