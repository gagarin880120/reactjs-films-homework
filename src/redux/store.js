import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';

const initialState = {
  searchResults: [],
  genres: [],
  isModalOpen: false,
  trailerURL: '',
  isTrailerLoaded: false,
};

export function reducer(state = initialState, action) {
  switch (action.type) {
    case 'SEARCH':
      return { ...state, searchResults: action.results };
    case 'GENRES':
      return { ...state, genres: action.genres };
    case 'MODAL':
      return { ...state, isModalOpen: action.isModalOpen };
    case 'TRAILER':
      return { ...state, trailerURL: action.trailerURL };
    case 'IS_TRAILER_LOADED':
      return { ...state, isTrailerLoaded: action.isTrailerLoaded };
    default:
      return state;
  }
}

export const store = createStore(reducer, applyMiddleware(thunk));
