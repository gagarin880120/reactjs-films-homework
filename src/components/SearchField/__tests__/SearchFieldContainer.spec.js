import ShallowRenderer from 'react-test-renderer/shallow';
import React from 'react';
import { SearchFieldContainer, mapDispatchToProps } from '../SearchFieldContainer';
import { getResults } from '../../../redux/actions';

jest.mock('../../../redux/actions',() => ({
  getResults: jest.fn().mockReturnValue('getResultsAction'),
}))

const renderComponent = (props = {}) => {
  const renderer = new ShallowRenderer();
  renderer.render(<SearchFieldContainer onSearch={jest.fn()} {...props} />);

  return renderer
}

describe('SearchFieldContainer', () => {
  describe('component', () => {
    test('SearchFieldContainer component should render without crashing', () => {
      const component = renderComponent()

      const result = component.getRenderOutput();

      expect(result).toMatchSnapshot();
    });

  })

  describe('mapDispatchToProps', () => {
    afterEach(() => {
      jest.clearAllMocks();
    })

    test('onSearch should dispatch getResults action', () => {
      const dispatch = jest.fn();
      const {onSearch} = mapDispatchToProps(dispatch);

      onSearch('abc')

      expect(getResults).toHaveBeenCalledWith('abc')
      expect(dispatch).toHaveBeenCalledWith('getResultsAction')
    })

    test('onSearch should not dispatch getResults action if no query', () => {
      const dispatch = jest.fn();
      const {onSearch} = mapDispatchToProps(dispatch);

      onSearch()

      expect(getResults).not.toHaveBeenCalled()
      expect(dispatch).not.toHaveBeenCalled()
    })
  })
})
