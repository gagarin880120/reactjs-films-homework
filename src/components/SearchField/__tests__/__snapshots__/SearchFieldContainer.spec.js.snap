// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchFieldContainer component SearchFieldContainer component should render without crashing 1`] = `
<SearchField
  onSearch={[MockFunction]}
/>
`;
